src/main/scala/io/picnicml/doddlemodel/preprocessing/QuantileDiscretizer.scala

```diff
@@ -0,0 +1,101 @@
+package io.picnicml.doddlemodel.preprocessing
+
+import breeze.linalg.DenseVector
+import breeze.stats.DescriptiveStats
+import io.picnicml.doddlemodel.CrossScalaCompat._
+import io.picnicml.doddlemodel.data.Feature.FeatureIndex
+import io.picnicml.doddlemodel.data.Features
+import io.picnicml.doddlemodel.syntax.OptionSyntax._
+import io.picnicml.doddlemodel.typeclasses.Transformer
+import scala.Double.{MinValue, MaxValue}
+
+case class QuantileDiscretizer(private val bucketCounts: DenseVector[Double], private val featureIndex: FeatureIndex, private val quantiles: Option[Seq[Seq[(Double, Double)]]] = None) {
+  private val numNumeric = featureIndex.numerical.columnIndices.length
+  require(numNumeric == 0 || numNumeric == bucketCounts.length, "A quantile should be given for every numerical column")
+}
+
+/** An immutable preprocessor that discretizes numerical features into buckets based on quantiles.
+  * Numerical feature values are converted to categorical features based on their distribution
+  * */
+object QuantileDiscretizer {
+
+  /** Create a quantile discretizer which splits data into discrete evenly sized buckets.
+    *
+    * @param bucketCounts The number of quantiles desired
+    * @param featureIndex feature index associated with features - this is needed so that only numerical features are
+    *                     transformed by this preprocessor; could be a subset of columns to be transformed
+    *
+    * @example Discretize a matrix into quartiles with two features: one numerical and one categorical.
+    *   {{{
+    *     import io.picnicml.doddlemodel.preprocessing.QuantileDiscretizer.ev
+    *
+    *     val featureIndex = FeatureIndex(List(NumericalFeature, CategoricalFeature))
+    *     val x = DenseMatrix(
+    *       List(-1.0, 0.0),
+    *       List(0.0, 1.0),
+    *       List(2.0, 0.0),
+    *       List(5.0, 0.0)
+    *     )
+    *     // equivalently, DenseVector(4) could be used
+    *     val bucketCounts = 4
+    *     val discretizer = QuantileDiscretizer(bucketCounts, featureIndex)
+    *     // Note: no fitting required
+    *     val xTransformed = ev.transform(discretizer, x)
+    *   }}}
+    */
+  def apply(bucketCount: Int, featureIndex: FeatureIndex): QuantileDiscretizer = {
+    val numNumeric = featureIndex.numerical.columnIndices.length
+    val bucketCountExtended: DenseVector[Double] = DenseVector.fill(numNumeric) { bucketCount.toDouble }
+    QuantileDiscretizer(bucketCountExtended, featureIndex)
+  }
+
+  implicit lazy val ev: Transformer[QuantileDiscretizer] = new Transformer[QuantileDiscretizer] {
+
+
+
+    @inline override def isFitted(model: QuantileDiscretizer): Boolean = model.quantiles.isDefined
+
+    override def fit(model: QuantileDiscretizer, x: Features): QuantileDiscretizer = {
+      val xCopy = x.copy
+      val discreteRangeArrays: Seq[Seq[(Double, Double)]] = model.featureIndex.numerical.columnIndices.zipWithIndex.map {
```

Comment:
No need for an additional explicit type here, i.e. `Seq[Seq[(Double, Double)]]` can be removed.

Reply:
Oops yeah, I explicitly set values during dev sometimes, will remove!